kerastuner/engine/oracle.py

```diff
@@ -17,90 +17,241 @@
+        self.tune_new_entries = tune_new_entries
+
+        # trial_id -> Trial
+        self.trials = {}
+        # tuner_id -> Trial
+        self.ongoing_trials = {}
+
+    def get_space(self):
+        return self.hyperparameters.copy()
+
+    def update_space(self, hyperparameters):
         """Add new hyperparameters to the tracking space.
 
         Already recorded parameters get ignored.
 
         Args:
-            new_entries: A list of HyperParameter objects to track.
+            hyperparameters: An updated HyperParameters object.
         """
-        ref_names = {p.name for p in self.space}
-        for p in new_entries:
-            if p.name not in ref_names:
-                self.space.append(p)
+        ref_names = {hp.name for hp in self.hyperparameters.space}
+        new_hps = [hp for hp in hyperparameters.space
+                   if hp.name not in ref_names]
 
-    def populate_space(self, trial_id, space):
-        """Fill a given hyperparameter space with values.
+        if new_hps and not self.allow_new_entries:
+            raise RuntimeError('`allow_new_entries` is `False`, but found '
+                               'new entries {}'.format(new_hps))
+
+        if not self.tune_new_entries:
+            # New entries should always use the default value.
+            return
+
+        for hp in new_hps:
+            self.hyperparameters.register(
+                hp.name, hp.__class__.__name__, hp.get_config())
+
+    def populate_space(self, trial_id):
+        """Fill the hyperparameter space with values for a trial.
 
         Args:
-            trial_id: String. Unique id for this trial.
-            space: A list of HyperParameter objects
-                to provide values for.
+          `trial_id`: The id for this Trial.
 
         Returns:
-            A dictionary with keys:
-                - status: string, one of "RUN", "IDLE", "EXIT"
-                - values: Only included if status is "RUN".
-                    Dict mapping parameter names to suggested values.
-                    Note that if the Oracle is keeping tracking of a large
-                    space, it may return values for more parameters
-                    than what was listed in `space`.
+            A dictionary with keys "values" and "status", where "values" is
+            a mapping of parameter names to suggested values, and "status"
+            is the TrialStatus that should be returned for this trial (one
+            of "RUNNING", "IDLE", or "STOPPED").
         """
         raise NotImplementedError
 
-    def result(self, trial_id, score):
-        """Record the measured objective for a set of parameter values.
+    def create_trial(self, tuner_id):
+        # Allow for multi-worker DistributionStrategy within a Trial.
+        if tuner_id in self.ongoing_trials:
+            return self.ongoing_trials[tuner_id]
 
-        If not overridden, this method does nothing.
+        trial_id = trial_lib.generate_trial_id()
```

Comment:
If an Oracle wants to generate a trial id, this will make it harder.
Can this be a member method with a default implementation calling
trial_lib.generate_trial_id()?

Reply:
I think for our use case we'll have to overwrite the entirety of `create_trial`